File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.py
import sys
from collections import defaultdict

input = sys.stdin.readline

k, n = map(int, input().split())
lens = [int(input()) for _ in range(k)]

def solution():
	# k개의 랜선을 가지고 있다.
	# k -> n개이상 같은 길이 랜선
	# 최대 랜선 길이?

	answer = 1
	s, e = 1, max(lens)

	while s <= e:
		mid = (s+e)//2

		cnt = 0

		for size in lens: cnt += size//mid

		if cnt >= n:
			answer = max(answer, mid)
			s = mid + 1
		else:
			e = mid - 1

	print(answer)
solution()

File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.js
const fs = require('fs');
const filePath = process.platform === 'linux' ? '/dev/stdin' : __dirname + '/input.txt';
const input = fs.readFileSync(filePath).toString().trim().split('\n');

const solution = () => {
  const [k, n] = input[0].split(' ').map(Number);
  const lens = new Array(k).fill(0).map((_, i) => Number(input[i + 1]));

  let [start, end] = [1, Math.max(...lens)];
  let answer = 0;

  while (start <= end) {
    let mid = parseInt((start + end) / 2);

    const count = lens.reduce((acc, val) => {
      return acc + parseInt(val / mid);
    }, 0);

    if (count >= n) {
      start = mid + 1;
      answer = mid;
    } else {
      end = mid - 1;
    }
  }

  console.log(answer);
};

solution();
